packages/zoe/src/contractSupport/recorder.js

```diff
@@ -0,0 +1,119 @@
+import { prepareDurablePublishKit } from '@agoric/notifier';
+import { M, prepareExoClass } from '@agoric/vat-data';
+import { E } from '@endo/eventual-send';
+
+/**
+ * Recorders support publishing data to vstorage.
+ *
+ * `Recorder` is similar to `Publisher` (in that they send out data) but has different signatures:
+ * - methods are async because they await remote calls to Marshaller and StorageNode
+ * - method names more aligned with [async iterators](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-iterator-interface)
+ * - omits fail()
+ * - adds getStorageNode() from its durable state
+ *
+ * Other names such as StoredPublisher or ChainStoragePublisher were considered, but found to be sometimes confused with *durability*, another trait of this class.
+ */
+
+/**
+ * @template T
+ * @typedef {{ getStorageNode: () => StorageNode, next(value: T): Promise<void>, return(value: T): Promise<void> }} Recorder
+ */
+
+/**
+ * @template T
+ * @typedef {Pick<PublishKit<T>, 'subscriber'> & { recorder: Recorder<T> }} RecorderKit
+ */
+
+/**
+ * @template T
+ * @typedef {Pick<PublishKit<T>, 'subscriber'> & { recorder: ERef<Recorder<T>> }} EventualRecorderKit
+ */
+
+/**
+ * Wrap a Publisher to record all the values to chain storage.
+ *
+ * @param {import('@agoric/zoe').Baggage} baggage
+ * @param {ERef<Marshaller>} marshaller
+ * @param {string} [kindName]
+ */
+export const prepareRecorder = (
+  baggage,
+  marshaller,
+  kindName = 'Chain Storage Publisher',
+) => {
```

Comment:
When a kind name argument exists for a prepare function, don't we conventionally require it and place it immediately after the baggage argument?
```suggestion
export const prepareRecorder = (baggage, kindName, marshaller) => {
```
```suggestion
export const prepareRecorder = (baggage, marshaller) => {
```

Reply:
fine by me

Reply:
oh I put kindName after marshaller because I wanted it to be optional. And the reason for that is I couldn't think of a reason the contract would be concerned with the kind name or want to have more than one per vat. Why is it important to have,
```js
const makeRecorder = prepareRecorder(baggage, 'Recorder', marshaller);
```
vs,
```js
const makeRecorder = prepareRecorder(baggage, marshaller);
```

Reply:
I guess I'd turn that around and ask why it's important to have the kind name be optional rather than either required or (more likely) non-configurable? But I don't consider any of those approaches to be necessary, so consider this as fully optional feedback.

Reply:
non-configurable is ideal to me. sounds like you also think that would improve this PR so I'll go with that. (there's only ever one baggage and one storage marshaller so no reason to define a second kind within the vat)